### PR TITLE
Huge refacto take #2

### PR DESCRIPTION
**Pull request by goldoraf**
This refactoring aims to simplify the app's code (obviously). Routing concerns have been isolated, getState() calls in action creators have been removed, some containers have been dumbified... Due to its size, this PR would probably been best reviewed by pulling the whole code instead of looking at the diffs.

Please note that there are still work to do: I'll (hopefully) add in the next hours some more commits, but the sooner this PR is merged, the better. So don't hesitate to have a look at it right now.

#### Checklist

Before merging this PR, the following things must have been done:

- [ ] Tested on supported browsers, including responsive mode (IE11 and Edge, Chrome >= 42, 2 latest versions of Firefox and Safari)
- [ ] All changes have test coverage

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -9,5 +9,5 @@
     }],
     "react"
   ],
-  "plugins": ["transform-object-rest-spread"]
+  "plugins": ["transform-object-rest-spread", "transform-class-properties"]
```
Comment (enguerran) [inline .babelrc]: I :heart: this (cf. https://babeljs.io/docs/plugins/transform-class-properties/)

File: src/reducers/view.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+export const getVisibleFiles = ({ view }) => {
+  const { files } = view
+  return getSortedFiles(files)
+}
+
+export const getVirtualRootFromUrl = pathname => {
+  if (pathname.match(/^\/files/)) return ROOT_DIR_ID
+  if (pathname.match(/^\/trash/)) return TRASH_DIR_ID
```
Comment (enguerran) [inline src/reducers/view.js]: You should use [`RegExp.prototype.test()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test) as it is way faster (see my previous comment https://github.com/cozy/cozy-files-v3/pull/205/commits/00fae34b1726daf53936e8b1799a8239db3014af#r108852976)

File: src/reducers/view.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+export const getUrlFromParams = ({ virtualRoot, displayedFolder }) => {
+  let path = virtualRoot === TRASH_DIR_ID ? '/trash' : '/files'
```
Comment (enguerran) [inline src/reducers/view.js]: This should be `const`.

File: src/actions/index.js [hunk 4]
```diff
@@ -45,28 +47,27 @@ const ALERT_LEVEL_ERROR = 'error'
...
 export const downloadFileOffline = () => ({ type: DOWNLOAD_FILE_E_OFFLINE, alert: { message: 'error.download_file.offline', level: ALERT_LEVEL_ERROR } })
 
-export const locationChange = (virtualRoot, folderId) => {
-  return async dispatch => {
-    dispatch({ type: LOCATION_CHANGE, virtualRoot, folderId })
-    return dispatch(openFolder(virtualRoot, folderId))
-  }
+export const openFiles = () => {
+  return async dispatch => dispatch(openFolder(ROOT_DIR_ID))
+}
```
Comment (enguerran) [inline src/actions/index.js]: Cannot we write it like:

```js
export const openFiles = () => async (dispatch) => dispatch(openFolder(ROOT_DIR_ID))
```

or (if we don't want parenthesis for single parameter in arrow function, see my previous comment https://github.com/cozy/cozy-files-v3/pull/205/commits/f67437f801b54a346ab00eac603568b02e15cc99#r108860052)

```js
export const openFiles = () => async dispatch => dispatch(openFolder(ROOT_DIR_ID))
```

File: src/components/FolderView.jsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  switch (fetchStatus) {
+    case 'pending':
+      return <Loading message={props.t('loading.message')} />
+    case 'failed':
+      return <Oops />
+    case 'loaded':
+      return files.length === 0 && !isAddingFolder
+        ? <Empty canUpload={canUpload} />
+        : <FileList {...props} />
+  }
```
Comment (enguerran) [inline src/components/FolderView.jsx]: We should add a `default` case for this `switch` like:

```js
default:
  return <Oops />
```

File: src/components/Topbar.jsx [hunk 6]
```diff
@@ -1,21 +1,21 @@
...
 
+// TODO: move the Topbar into FolderView so that we can get rid of this
 import { ROOT_DIR_ID, TRASH_DIR_ID } from '../constants/config'
-
-const TOOLBARS = {
-  [ROOT_DIR_ID]: FilesToolbar,
-  [TRASH_DIR_ID]: TrashToolbar
+const getToolbar = location => {
+  if (location.pathname.match(/^\/files/)) return FilesToolbar
+  if (location.pathname.match(/^\/trash/)) return TrashToolbar
```
Comment (enguerran) [inline src/components/Topbar.jsx]: RegExp.prototype.test() is faster than String.prototype.match.

**Comment by enguerran** [comment]
Broadly speaking I see very good changes in the code architecture with the use of selectors, the decoupling of things and the SRP.

File: src/actions/index.js [hunk 7]
```diff
@@ -46,25 +45,28 @@ const ALERT_LEVEL_ERROR = 'error'
...
+
+export const openFolder = (virtualRoot, folderId) => {
+  return async dispatch => {
+    if (!folderId) folderId = virtualRoot
+    dispatch({ type: OPEN_FOLDER, virtualRoot, folderId })
     let folder, parent
     try {
       folder = await cozy.client.files.statById(folderId)
       const parentId = folder.attributes.dir_id
       parent = !!parentId && await cozy.client.files.statById(parentId)
```
Comment (goldoraf) [inline src/actions/index.js]: Right, done.

File: src/actions/index.js [hunk 8]
```diff
@@ -74,9 +76,8 @@ export const openFolder = (folderId, context = FILES_CONTEXT) => {
   }
 }
 
-export const openFileInNewTab = (file) => {
-  return async (dispatch, getState) => {
-    const folder = getState().folder
+export const openFileInNewTab = (folder, file) => {
+  return async dispatch => {
     // TODO: replace this with cozy.client.getFilePath(file, folder)
```
Comment (goldoraf) [inline src/actions/index.js]: I did the replacement, thanks for the reminder ;)

File: src/actions/index.js [hunk 9]
```diff
@@ -118,15 +119,9 @@ export const abortAddFolder = (accidental) => {
   return action
 }
 
-export const renameFolder = (newName, id) => ({
-  type: RENAME_FOLDER,
-  id,
-  name: newName
-})
-
 export const createFolder = name => {
```
Comment (goldoraf) [inline src/actions/index.js]: Yes, but I prefer to wait for the moment we'll add tests to that part.

File: src/actions/index.js [hunk 10]
```diff
@@ -148,7 +143,7 @@ export const createFolder = name => {
     try {
       folder = await cozy.client.files.createDirectory({
         name: name,
-        dirID: getState().folder.id
+        dirID: getState().view.displayedFolder.id
```
Comment (goldoraf) [inline src/actions/index.js]: yes, there still are places to refactor in order to get rid of these `getState()` antipatterns

File: src/components/FolderView.jsx (hunk 5 above)
Comment (goldoraf) [inline src/components/FolderView.jsx]: Done

File: src/containers/Nav.jsx [hunk 11]
```diff
@@ -21,27 +22,41 @@ class ActiveLink extends Component {
...
+        this.props.router.push(this.props.to)
+      })
   }
 
-  render ({ to, className, activeClassName, children }, { opening }) {
+  render () {
+    const { router, location, to, activeClassName, children, ...props } = this.props
+    const { opening } = this.state
+
+    props.href = router.createHref(to)
```
Comment (goldoraf) [inline src/containers/Nav.jsx]: I needed a different behavior than the standard's `Link` (the `setState` call that displays the spinner)

File: src/components/File.jsx [hunk 12]
```diff
@@ -5,7 +5,8 @@ import { withRouter } from 'react-router'
 
 import styles from '../styles/table'
 import { translate } from '../lib/I18n'
-import FilenameInput from '../components/FilenameInput'
+
+import { getUrlFromParams } from '../reducers'
```
Comment (goldoraf) [inline src/components/File.jsx]: This one is not even a selector, more a helper

File: src/containers/FileExplorer.jsx [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+const urlHasChanged = (props, newProps) =>
+  props.location.pathname !== newProps.location.pathname
+
+const matchCurrentView = (props, newProps) =>
+  newProps.location.pathname === getUrlFromParams(props.view)
+
+class FileExplorer extends Component {
+  componentWillMount () {
+    this.props.onLocationChange(
+      getVirtualRootFromUrl(this.props.location.pathname),
```
Comment (goldoraf) [inline src/containers/FileExplorer.jsx]: Not sure ;)

File: src/containers/Folder.jsx [hunk 14]
```diff
@@ -23,84 +23,64 @@ import TrashSelectionBar from '../containers/TrashSelectionBar'
...
 const mapStateToProps = (state, ownProps) => ({
-  visibleContext: state.context,
-  isFetching: state.ui.isFetching,
-  showLoading: state.ui.isFetching && state.folder === null,
-  error: state.ui.error,
+  virtualRoot: state.view.virtualRoot,
+  displayedFolder: state.view.displayedFolder,
+  fetchStatus: state.view.fetchStatus,
+  // error: state.ui.error,
+  isAddingFolder: mustShowAddFolder(state), // not fan of this...
```
Comment (goldoraf) [inline src/containers/Folder.jsx]: It's gone ;)

File: src/containers/Nav.jsx [hunk 15]
```diff
@@ -26,17 +26,19 @@ class ActiveLink extends Component {
...
         {opening && <Spinner />}
       </Link>
     )
   }
 }
 
-const Nav = ({ t, context, openFiles, openTrash }) => {
+const Nav = ({ t, location, openFiles, openTrash }) => {
+  const isBrowsingFiles = location.pathname.match(/^\/files/) !== null
+  const isBrowsingTrash = location.pathname.match(/^\/trash/) !== null
```
Comment (goldoraf) [inline src/containers/Nav.jsx]: It's faster but less readable I think...

File: src/actions/index.js (hunk 9 above)
Comment (enguerran) [inline src/actions/index.js]: LVGTM (Look Very Good To Me)

File: src/containers/Nav.jsx (hunk 11 above)
Comment (enguerran) [inline src/containers/Nav.jsx]: okay, understood

**Comment by enguerran** [comment]
:v:

**Pull request merged**